pkg/proxy/ipvs/graceful_delete.go

```diff
@@ -0,0 +1,250 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipvs
+
+import (
+	"sync"
+	"time"
+
+	"container/heap"
+	"fmt"
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	utilipvs "k8s.io/kubernetes/pkg/util/ipvs"
+)
+
+const (
+	RSGracefulDeletePeriod = 15 * time.Minute
+	RSCheckDeleteInterval  = 1 * time.Minute
+)
+
+// GracefulDeleteRS stores real server information and the process time.
+// If nothing special happened, real server will be delete after process time.
+type GracefulDeleteRS struct {
+	VirtualServer *utilipvs.VirtualServer
+	RealServer    *utilipvs.RealServer
+	ProcessAt     time.Time
+}
+
+// String return the unique real server name(with virtual server information)
+func (g *GracefulDeleteRS) String() string {
+	return g.VirtualServer.String() + "/" + g.RealServer.String()
+}
+
+// GracefulDeleteRSQueueQueue is a priority heap where the lowest ProcessAt is at the front of the queue
+type GracefulDeleteRSQueueQueue []*GracefulDeleteRS
```

Comment:
Just an idea: why not make the internal type of the heap a map with the key being the queue item and the value being the index? That way you can use the native properties of the map to preserve uniqueness and you should still be able to implement the sort and heap interfaces. Then, you won't need the UniqueQueue .Maybe it does not make sense but just wanted to throw that idea out there.